File: codex/src/domain/entities/usuario.ts
import { Pedido } from "./pedido.js";
import { PapelUsuario } from "../enums/papel-usuario.js";
import { DomainError } from "../errors/domain-error.js";

export type UsuarioProps = {
  id: number;
  nome: string;
  email: string;
  senhaHash: string;
  papel: PapelUsuario;
  dataCadastro?: Date;
};

export class Usuario {
  readonly id: number;
  nome: string;
  email: string;
  private readonly senhaHash: string;
  papel: PapelUsuario;
  readonly dataCadastro: Date;

  constructor(props: UsuarioProps) {
    if (!props.nome.trim()) throw new DomainError("Nome do usuario e obrigatorio.");
    if (!props.email.trim()) throw new DomainError("Email do usuario e obrigatorio.");
    if (!props.senhaHash.trim()) throw new DomainError("Hash da senha e obrigatorio.");

    this.id = props.id;
    this.nome = props.nome;
    this.email = props.email;
    this.senhaHash = props.senhaHash;
    this.papel = props.papel;
    this.dataCadastro = props.dataCadastro ?? new Date();
  }

  verificarSenha(senhaHash: string): boolean {
    return this.senhaHash === senhaHash;
  }

  obterSenhaHash(): string {
    return this.senhaHash;
  }

  ehAdministrador(): boolean {
    return this.papel === PapelUsuario.ADMIN;
  }

  podeVisualizarPedido(pedido: Pedido): boolean {
    return this.ehAdministrador() || pedido.usuarioId === this.id;
  }
}

File: claude/backend/src/main/java/com/ecommerce/domain/Usuario.java
package com.ecommerce.domain;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import java.time.LocalDateTime;

/**
 * Usuario do sistema: representa clientes e administradores (diagrama de dominio).
 *
 * <p>Regras: {@code email} unico; a senha e armazenada apenas como hash em
 * {@code senhaHash} (RF05), nunca em texto puro.</p>
 */
@Entity
@Table(name = "usuario")
public class Usuario {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String nome;

    @Column(nullable = false, unique = true)
    private String email;

    @Column(name = "senha_hash", nullable = false)
    private String senhaHash;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private PapelUsuario papel;

    @Column(name = "data_cadastro", nullable = false)
    private LocalDateTime dataCadastro;

    protected Usuario() {
        // exigido pelo JPA
    }

    public Usuario(String nome, String email, String senhaHash, PapelUsuario papel) {
        this.nome = nome;
        this.email = email;
        this.senhaHash = senhaHash;
        this.papel = papel;
        this.dataCadastro = LocalDateTime.now();
    }

    /** Verifica a senha em texto puro contra o hash armazenado (RF05). */
    public boolean verificarSenha(String senhaPlana, EncoderSenha encoder) {
        return encoder.corresponde(senhaPlana, this.senhaHash);
    }

    public boolean ehAdministrador() {
        return this.papel == PapelUsuario.ADMIN;
    }

    /** RN05: apenas o proprio cliente ou um administrador pode visualizar um pedido. */
    public boolean podeVisualizarPedido(Pedido pedido) {
        return ehAdministrador() || (pedido != null && this.id != null && this.id.equals(pedido.getUsuarioId()));
    }

    public Long getId() {
        return id;
    }

    public String getNome() {
        return nome;
    }

    public String getEmail() {
        return email;
    }

    public String getSenhaHash() {
        return senhaHash;
    }

    public PapelUsuario getPapel() {
        return papel;
    }

    public LocalDateTime getDataCadastro() {
        return dataCadastro;
    }
}
